src/rwhod/daemon.3
.\"	From: NetBSD: daemon.3,v 1.3 1995/02/25 13:41:12 cgd Exp
.\"	$Id: daemon.3,v 1.12 2000/07/30 23:57:06 dholland Exp $
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)daemon.3	8.1 (Berkeley) 6/9/93
.\"
.Dd June 9, 1993
.Dt DAEMON 3
.Os "Linux NetKit (0.17)"
.Sh NAME
.Nm daemon
.Nd run in the background
.Sh SYNOPSIS
.Fd #include <stdlib.h>
.Fn daemon "int nochdir" "int noclose"
.Sh DESCRIPTION
.Pp
The
.Fn daemon
function is for programs wishing to detach themselves from the
controlling terminal and run in the background as system daemons.
.Pp
Unless the argument
.Fa nochdir
is non-zero,
.Fn daemon
changes the current working directory to the root (``/'').
.Pp
Unless the argument
.Fa noclose
is non-zero,
.Fn daemon
will redirect standard input, standard output and standard error
to ``/dev/null''.
.Sh ERRORS
The function
.Fn daemon
may fail and set
.Va errno
for any of the errors specified for the library functions
.Xr fork 2
and 
.Xr setsid 2 .
.Sh SEE ALSO
.Xr setsid 2
.Sh HISTORY
The
.Fn daemon
function first appeared in
.Bx 4.4 .


src/ChangeLog
22-Jul-2000:
	Add some minor additional paranoia.
	Changes from Olaf Kirch to run chrooted to /var/spool/rwho and
	  then optionally drop root privileges.

3-Jun-2000:
	Fix sizes of types in enclosed protocols/rwhod.h.

14-Dec-1999:
	netkit-rwho-0.16 is released.

1-Aug-1999:
	Complete y2k and y2038 audit. Clean up y2038 issues.
	Clean up buffer problems in rwho (apparently unexploitable, 
	  fortunately.)

31-Jul-1999:
	Redid makefiles/config stuff for new confgen version.

23-Sep-1997:
	Added ruptime.
	Fixed a suspicious strncpy in rwhod.

12-Jun-1997:
	netkit-rwho-0.10 released.

08-Jun-1997:
	More adjustments for glibc.

19-May-1997:
	glibc fixes from Red Hat.

13-May-1997:
	Patch from Debian to support selective use of interfaces in 
	rwhod. (via Peter Tobias, tobias@et-inf.fho-emden.de)

05-Apr-1997:
	Added configure script to generate MCONFIG.
	Ensure null termination of certain input data. (Peter Tobias,
	  tobias@et-inf.fho-emden.de)
	Better utmp handling in rwhod.

08-Mar-1997: 
	Split from full NetKit package. 
	Generated this change log from NetKit's.

29-Dec-1996
	NetKit-0.09 released.
	Assorted alpha/glibc patches. (Erik Troan, ewt@redhat.com)
	Assorted bug fixes from Debian. (Peter Tobias, 
	  tobias@et-inf.fho-emden.de)
	Hardened programs against DNS h_length spoofing attacks.
	Use inet_aton() everywhere instead of inet_addr().

22-Aug-1996
	NetKit-B-0.08 released.
	(almost) everything now compiles with lots of warnings turned on.
	Fixed rwhod not to need -lbsd.
	Fixed possible buffer overrun hole in rwhod.

25-Jul-1996
	NetKit-B-0.07A released.

23-Jul-1996
	NetKit-B-0.07 released.
	Integrated a collection of patches that had been lurking on the net,
	  including the 256-ptys support for telnetd and passive mode ftp.
	Major security fixes, including to fingerd, lpr, rlogin, rsh, talkd, 
	  and telnetd. Do *not* use the sliplogin from earlier versions of this
	  package, either.
	Much of the code builds without libbsd.a or bsd includes.
	Massive code cleanup. Almost everything compiles clean with gcc
	  -Wall now. rusers and rusersd do not; patches to rpcgen to fix
	  this would be appreciated if anyone feels like it.
	New maintainer:  David A. Holland, dholland@hcs.harvard.edu

date not known
	NetKit-B-0.06 released.

date not known
	NetKit-B-0.05 released.

date not known
	NetKit-B-0.04 released.

date not known
	NetKit-B-0.03 released.
	Changed rusers, rpc.ruserd, rwall, rpc.rwalld, rwho, rwhod to not
	  require to copy ./include/*.h files into /usr/include



src/rwhod/daemon.c
/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * From: @(#)daemon.c	8.1 (Berkeley) 6/4/93
 *       NetBSD: daemon.c,v 1.4 1995/02/25 13:41:16 cgd Exp
 */
char daemon_rcsid[] = 
  "$Id: daemon.c,v 1.2 1996/08/15 07:22:28 dholland Exp $";

#include <fcntl.h>
#include <paths.h>
#include <unistd.h>
#include "daemon.h"

int daemon(int nochdir, int noclose)
{
	int fd;

	switch (fork()) {
	  case -1:
		return -1;
	  case 0:
		break;
	  default:
		_exit(0);
	}

	if (setsid() == -1) return -1;
	if (!nochdir) chdir("/");
	if (noclose) return 0;

	fd = open(_PATH_DEVNULL, O_RDWR, 0);
	if (fd != -1) {
		dup2(fd, STDIN_FILENO);
		dup2(fd, STDOUT_FILENO);
		dup2(fd, STDERR_FILENO);
		if (fd > 2) close(fd);
	}
	return 0;
}
